pkg/repository/git.go

```diff
@@ -122,53 +120,60 @@ func (r gitRepoController) getSHA() (string, error) {
 	return head.Hash().String(), nil
 }
 
-func (r gitRepoController) getTreeSHA(context string) (string, error) {
-	repo, err := r.repoGetter.get(r.path)
-	if err != nil {
-		return "", fmt.Errorf("failed to get repository: %w", err)
-	}
+type commitResponse struct {
+	commit string
+	err    error
+}
 
-	ref, err := repo.Head()
-	if err != nil {
-		return "", fmt.Errorf("failed to get HEAD from repo: %w", err)
-	}
+func (r gitRepoController) GeLatestDirCommit(contextDir string) (string, error) {
+	// We use context.TODO() in a few places to call isClean, so let's make sure
+	// we set proper internal timeouts to not leak goroutines
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
 
-	commit, err := repo.CommitObject(ref.Hash())
-	if err != nil {
-		return "", fmt.Errorf("failed to get commit object from reference: %w", err)
-	}
-	tree, err := commit.Tree()
-	if err != nil {
-		return "", fmt.Errorf("failed to get tree from commit: %w", err)
-	}
+	timeoutCh := make(chan struct{})
+	ch := make(chan commitResponse)
 
-	if context == "." {
-		return tree.Hash.String(), nil
-	}
+	timeoutErr := errors.New("timeout exceeded")
+	timeout := 3 * time.Second
 
-	treeDir, err := getTreeDirFromRelPath(context)
-	if err != nil {
-		oktetoLog.Infof("could not tree working dir: %w", err)
-		return "", err
-	}
+	go func() {
+		time.Sleep(timeout)
+		close(timeoutCh)
+		cancel()
+		ch <- commitResponse{
+			commit: "",
+			err:    timeoutErr,
+		}
+	}()
 
-	svcEntry, err := tree.FindEntry(treeDir)
-	if err != nil {
-		return "", fmt.Errorf("failed to find an entry in tree: %w", err)
+	go func() {
+		commit, err := r.calculateLatestDirCommit(ctx, contextDir)
+		select {
+		case <-timeoutCh:
+		case ch <- commitResponse{
+			commit: commit,
+			err:    err,
+		}:
+		}
+	}()
+
+	s := <-ch
+
+	if s.err == timeoutErr {
+		oktetoLog.Debugf("Timeout exceeded calculating git commit for '%s': assuming dirty commit", contextDir)
 	}
 
-	return svcEntry.Hash.String(), nil
+	return s.commit, s.err
 }
 
-func getTreeDirFromRelPath(relPath string) (string, error) {
-
-	wdCtrl := filesystem.NewOsWorkingDirectoryCtrl()
-	wd, err := wdCtrl.Get()
+func (r gitRepoController) calculateLatestDirCommit(ctx context.Context, contextDir string) (string, error) {
+	repo, err := r.repoGetter.get(r.path)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to analyze git repo: %w", err)
```

Comment:
shall we have a more specific error string here? I suggest this because we already have a few errors that say `failed to analyze git repo` and I am afraid it will become difficult to identify quickly from which of these errors the error is coming from.